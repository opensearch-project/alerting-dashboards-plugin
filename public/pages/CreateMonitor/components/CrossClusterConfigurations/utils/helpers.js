/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import { DEFAULT_EMPTY_DATA, MONITOR_TYPE } from '../../../../../utils/constants';

export const getLocalClusterName = async (httpClient) => {
  let localClusterName = DEFAULT_EMPTY_DATA;
  try {
    const response = await httpClient.get('../api/alerting/_health');
    if (response.ok) {
      localClusterName = response.resp[0]?.cluster;
    } else {
      console.log('Error getting local cluster name:', response);
    }
  } catch (e) {
    console.error(e);
  }
  return localClusterName;
};

export const getDataSources = (monitor, localClusterName) => {
  const monitorType = _.get(
    monitor,
    'monitor_type',
    _.get(monitor, 'ui_metadata.monitor_type', MONITOR_TYPE.QUERY_LEVEL)
  );
  let dataSources;
  switch (monitorType) {
    case MONITOR_TYPE.CLUSTER_METRICS:
      dataSources = _.get(monitor, 'inputs.0.uri.clusters');
      // To preserve functionality of legacy monitors, cluster metrics monitors run on the
      // local cluster by default if no clusters are specified in the monitor configuration.
      if (_.isEmpty(dataSources)) dataSources = [localClusterName || DEFAULT_EMPTY_DATA];
      break;
    case MONITOR_TYPE.DOC_LEVEL:
      dataSources = _.get(monitor, 'inputs.0.doc_level_input.indices', [DEFAULT_EMPTY_DATA]);
      break;
    default:
      dataSources = _.get(monitor, 'inputs.0.search.indices', [DEFAULT_EMPTY_DATA]);
  }
  dataSources = _.sortBy(dataSources);
  return dataSources;
};
