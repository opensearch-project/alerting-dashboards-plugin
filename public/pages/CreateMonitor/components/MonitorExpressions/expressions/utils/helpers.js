/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export function selectOptionValueToText(optionValue, options) {
  return options.find((opt) => opt.value === optionValue).text;
}

export function getOfExpressionAllowedTypes(values) {
  const types = ['number'];
  if (['min', 'max'].includes(values.aggregationType)) types.push('date');
  return types;
}

export function getMetricExpressionAllowedTypes(values) {
  const types = ['number'];

  // Check if the aggregation types includes any of 'min', 'max', or 'count'.
  // If so, add 'date' to the allowed type for metric expression field.
  if (
    values.aggregations.some((e) => {
      return ['min', 'max', 'count'].includes(e.aggregationType);
    })
  )
    types.push('date');

  // Check if the aggregations types includes 'count'.
  // If so, add 'keyword' to the allowed type for metric expression field.
  if (
    values.aggregations.some((e) => {
      return ['count'].includes(e.aggregationType);
    })
  )
    types.push('keyword');
  return types;
}

export function getGroupByExpressionAllowedTypes() {
  return ['keyword'];
}

export const validateAggregationsDuplicates = (aggregations) => {
  let duplicates = 0;
  aggregations.forEach((e1, index) => {
    aggregations.slice(index + 1).forEach((e2) => {
      if (e1.aggregationType === e2.aggregationType && e1.fieldName === e2.fieldName) {
        duplicates++;
      }
    });
  });
  return duplicates > 0;
};
