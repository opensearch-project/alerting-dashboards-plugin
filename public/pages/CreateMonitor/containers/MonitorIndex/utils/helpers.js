/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export function canAppendWildcard(keyPressed) {
  // If it's not a letter, number or is something longer, reject it
  if (!keyPressed || !/[a-z0-9]/i.test(keyPressed) || keyPressed.length !== 1) {
    return false;
  }
  return true;
}

export function createReasonableWait(cb) {
  return setTimeout(cb, 100);
}

const MAX_NUMBER_OF_MATCHING_INDICES = 100;

export function filterSystemIndices(indices, isIncludingSystemIndices) {
  if (!indices) {
    return indices;
  }

  const acceptableIndices = isIncludingSystemIndices
    ? indices
    : // All system indices begin with a period.
      indices.filter((index) => !index.label.startsWith('.'));

  return acceptableIndices.slice(0, MAX_NUMBER_OF_MATCHING_INDICES);
}

/**
 This utility is designed to do a couple of things:

 1) Take in list of indices and filter out system indices if necessary
 2) Return a `visible` list based on a priority order.

 We are passing in three separate lists because they each represent
 something slightly different.

 - `unfilteredAllIndices`
 This is the result of the initial `*` query and represents all known indices
 - `unfilteredPartialMatchedIndices`
 This is the result of searching against the query with an added `*`. This is only
 used when the query does not end in an `*` and represents potential matches in the UI
 - `unfilteredExactMatchedIndices
 This is the result of searching against a query that already ends in `*`.
 We call this `exact` matches because ES is telling us exactly what it matches
 */
export function getMatchedOptions(
  unfilteredAllIndices,
  unfilteredPartialMatchedIndices,
  unfilteredExactMatchedIndices,
  unfilteredAllAliases,
  unfilteredPartialMatchedAliases,
  unfilteredExactMatchedAliases,
  isIncludingSystemIndices
) {
  const allIndices = filterSystemIndices(unfilteredAllIndices, isIncludingSystemIndices);
  const partialMatchedIndices = filterSystemIndices(
    unfilteredPartialMatchedIndices,
    isIncludingSystemIndices
  );
  const exactMatchedIndices = filterSystemIndices(
    unfilteredExactMatchedIndices,
    isIncludingSystemIndices
  );
  const allAliases = filterSystemIndices(unfilteredAllAliases, isIncludingSystemIndices);
  const partialMatchedAliases = filterSystemIndices(
    unfilteredPartialMatchedAliases,
    isIncludingSystemIndices
  );
  const exactMatchedAliases = filterSystemIndices(
    unfilteredExactMatchedAliases,
    isIncludingSystemIndices
  );
  // We need to pick one to show in the UI and there is a priority here
  // 1) If there are exact matches, show those as the query is good to go
  // 2) If there are no exact matches, but there are partial matches,
  // show the partial matches
  // 3) If there are no exact or partial matches, just show all indices
  let visibleOptions = [];
  if (exactMatchedAliases.length || exactMatchedIndices.length) {
    visibleOptions = [
      { label: 'Aliases', options: exactMatchedAliases },
      { label: 'Indices', options: exactMatchedIndices },
    ];
    // visibleOptions = exactMatchedAliases.concat(exactMatchedIndices);
  } else if (partialMatchedAliases.length || partialMatchedIndices.length) {
    visibleOptions = [
      { label: 'Aliases', options: partialMatchedAliases },
      { label: 'Indices', options: partialMatchedIndices },
    ];
    // visibleOptions = partialMatchedAliases.concat(partialMatchedIndices);
  } else {
    visibleOptions = [
      { label: 'Aliases', options: allAliases },
      { label: 'Indices', options: allIndices },
    ];
    // visibleOptions = allAliases.concat(allIndices);
  }

  return {
    visibleOptions,
  };
}
