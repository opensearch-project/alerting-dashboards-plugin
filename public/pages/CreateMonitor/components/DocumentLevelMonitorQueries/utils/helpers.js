/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import { SUPPORTED_DOC_LEVEL_QUERY_OPERATORS } from './constants';
import { OPERATORS_MAP } from '../../MonitorExpressions/expressions/utils/constants';
import { FORMIK_INITIAL_DOCUMENT_LEVEL_QUERY_VALUES } from '../../../containers/CreateMonitor/utils/constants';

/**
 * Returns an array of { text: string, value: string, dataTypes: string[] } objects.
 * Used by the DocumentLevelQuery.js to populate the operator options FormikSelect element.
 * @return {*[]}
 */
export const getDocLevelQueryOperators = () => {
  return SUPPORTED_DOC_LEVEL_QUERY_OPERATORS.map((type) => OPERATORS_MAP[_.toUpper(type)]);
};

/**
 * Validates whether the document level monitor query can be executed.
 * @param query - a FORMIK_INITIAL_DOCUMENT_LEVEL_QUERY_VALUES object.
 * @return {boolean} - TRUE if the query is valid; otherwise FALSE.
 */
export const validateDocLevelGraphQuery = (query = FORMIK_INITIAL_DOCUMENT_LEVEL_QUERY_VALUES) => {
  // The 'queryName', 'field', 'operator', and 'query' fields are required to execute a doc level query.
  // If any of those fields are undefined for any queries, the monitor cannot be executed.
  return (
    !_.isEmpty(query.queryName) &&
    !_.isEmpty(query.field) &&
    !_.isEmpty(query.operator) &&
    !_.isEmpty(query.query?.toString())
  );
};

/**
 * Validates whether the array of document level monitor queries can be executed.
 * @param queries - an array of FORMIK_INITIAL_DOCUMENT_LEVEL_QUERY_VALUES objects.
 * @return {boolean} - TRUE if there is at least 1 defined query, and all queries are valid; otherwise FALSE.
 */
export const validDocLevelGraphQueries = (queries = []) => {
  return !_.isEmpty(queries) && !queries.some((query) => !validateDocLevelGraphQuery(query));
};
