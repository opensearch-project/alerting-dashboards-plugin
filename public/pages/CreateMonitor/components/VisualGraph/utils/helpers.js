/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import moment from 'moment';

import { selectOptionValueToText } from '../../MonitorExpressions/expressions/utils/helpers';
import {
  AGGREGATION_TYPES,
  UNITS_OF_TIME,
} from '../../MonitorExpressions/expressions/utils/constants';
import {
  Y_DOMAIN_BUFFER,
  DEFAULT_MARK_SIZE,
  X_DOMAIN_BUFFER,
  BAR_PERCENTAGE,
  BAR_KEY_COUNT,
} from './constants';
import { MONITOR_TYPE } from '../../../../../utils/constants';

export function getYTitle(values) {
  return _.get(values, `aggregations[0].fieldName`, 'doc_count');
}

export function getLeftPadding(yDomain) {
  const [min, max] = yDomain;
  const maxLength = Math.max(min.toString().length, max.toString().length);
  // These are MAGIC numbers..
  // 80 is the default left padding
  // 9 is the length when we start adding to base
  // 10 is the number we multiply to add left padding for each digit
  const multiplier = Math.max(maxLength - 9, 0);
  return 80 + multiplier * 10;
}

export function getXDomain(data) {
  const minDate = data[0];
  const maxDate = data[data.length - 1];
  return [minDate.x, maxDate.x];
}

export function getBufferedXDomain(data, values) {
  const { bucketValue, bucketUnitOfTime } = values;
  const minDate = data[0].x;
  const maxDate = data[data.length - 1].x;
  // If minDate equals to maxDate, then use bucketValue and bucketUnitOfTime as timeRange.
  let timeRange = maxDate - minDate;
  if (!timeRange) timeRange = moment.duration(bucketValue, bucketUnitOfTime);

  const minDateBuffer = minDate - timeRange * X_DOMAIN_BUFFER;
  const maxDateBuffer = maxDate.getTime() + timeRange * X_DOMAIN_BUFFER;
  return [minDateBuffer, maxDateBuffer];
}

export function getYDomain(data) {
  if (!data.length) return [0, 10];
  const max = data.reduce((accu, { y }) => Math.max(accu, y), 0);
  const min = data.reduce((accu, { y }) => Math.min(accu, y), 0);
  if (max === min) return [0, 10];
  const maxBuffer = Math.ceil(max * Y_DOMAIN_BUFFER);
  const minBuffer = Math.floor(min * Y_DOMAIN_BUFFER);
  return [minBuffer, maxBuffer];
}

export function formatYAxisTick(value) {
  return value.toLocaleString();
}

export function getAnnotationData(xDomain, yDomain, thresholdValue) {
  const [xMin, xMax] = xDomain;
  const [yMin, yMax] = yDomain;
  let yValue = thresholdValue;
  if (thresholdValue > yMax) yValue = yMax;
  if (thresholdValue < yMin) yValue = yMin;
  return [
    { x: xMin, y: yValue },
    { x: xMax, y: yValue },
  ];
}

export function getDataFromResponse(response, fieldName, monitorType) {
  if (!response) return [];
  const isQueryMonitor = monitorType === MONITOR_TYPE.QUERY_LEVEL;

  if (isQueryMonitor) {
    const buckets = _.get(response, 'aggregations.over.buckets', []);
    return buckets.map(getXYValues).filter(filterInvalidYValues);
  } else {
    const buckets = _.get(response, 'aggregations.composite_agg.buckets', []);
    return buckets
      .map((bucket) => getXYValuesByFieldName(bucket, fieldName))
      .filter(filterInvalidYValues)
      .sort((a, b) => a.x - b.x);
  }
}

// Function for aggregation type monitors to get Map of data.
// The current response gives a large number of data aggregated in buckets, and this function returns the top n results with highest count of data points.
// The number n is based on the constant BAY_KEY_COUNT.
export function getMapDataFromResponse(response, fieldName, groupByFields) {
  if (!response) return [];
  const buckets = _.get(response, 'aggregations.composite_agg.buckets', []);
  const allData = new Map();
  buckets.map((bucket) => {
    const dataPoint = getXYValuesByFieldName(bucket, fieldName);
    // Key of object is the string concat by group by field values
    const key = groupByFields.map((field) => _.get(bucket.key, field, '-')).join(', ');
    allData.has(key)
      ? allData.set(key, [dataPoint, ...allData.get(key)])
      : allData.set(key, [dataPoint]);
  });
  const entryLength = [];
  for (const [key, value] of allData.entries()) {
    allData.set(key, _.filter(value, filterInvalidYValues));
    entryLength.push({ key, length: value.length });
  }
  // Return arrays of data with more data points
  entryLength.sort((entryA, entryB) => entryB.length - entryA.length);
  const result = entryLength.slice(0, BAR_KEY_COUNT).map((entry) => {
    return { key: entry.key, data: allData.get(entry.key) };
  });
  return result;
}

export function getXYValuesByFieldName(bucket, fieldName) {
  const x = new Date(bucket.key.date);
  // Parse the fieldName containing "." to "_"
  const parsedFieldName = fieldName.replace(/\./g, '_');
  const path = bucket[parsedFieldName] ? `${parsedFieldName}.value` : 'doc_count';
  const y = _.get(bucket, path, null);
  return { x, y };
}

export function getXYValues(bucket) {
  const x = new Date(bucket.key_as_string);
  const path = bucket.metric ? 'metric.value' : 'doc_count';
  const y = _.get(bucket, path, null);
  return { x, y };
}

export function filterInvalidYValues({ y }) {
  return !isNaN(parseFloat(y));
}

export function getMarkData(data) {
  return data.map((d) => ({ ...d, size: DEFAULT_MARK_SIZE }));
}

export function getRectData(data, width = 30000, index, seriesCount) {
  // Shift x and x0 value according to total number of data series
  const midIndex = seriesCount / 2;
  const shiftAmount = (index - midIndex) * width;
  return data.map((d) => {
    const x = d.x.getTime() + shiftAmount;
    const x0 = x - width;
    return {
      x0,
      x,
      y: d.y,
      actualX: d.x,
    };
  });
}

export function computeBarWidth(xDomain) {
  const [min, max] = xDomain;
  return Math.abs(max - min) * BAR_PERCENTAGE;
}

export function getAggregationGraphHint(hint) {
  return (
    hint.seriesName +
    ': (' +
    hint.data.actualX.toLocaleString() +
    ', ' +
    hint.data.y.toLocaleString() +
    ')'
  );
}

export function getGraphDescription(bucketValue, bucketUnitOfTime, groupBy) {
  const text = `FOR THE LAST ${bucketValue} ${selectOptionValueToText(
    bucketUnitOfTime,
    UNITS_OF_TIME
  )}`;
  if (_.isEmpty(groupBy)) return text;
  return text + ` GROUP BY ${groupBy}`;
}
