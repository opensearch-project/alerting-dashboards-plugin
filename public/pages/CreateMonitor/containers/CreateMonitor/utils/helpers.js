/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import queryString from 'query-string';
import { FORMIK_INITIAL_VALUES } from './constants';
import monitorToFormik from './monitorToFormik';
import { formikToMonitor } from './formikToMonitor';
import { MONITOR_TYPE } from '../../../../../utils/constants';
import { initializeFromQueryParams } from './monitorQueryParams';
import { backendErrorNotification, getDigitId } from '../../../../../utils/helpers';
import {
  formikToTrigger,
  formikToTriggerUiMetadata,
} from '../../../../CreateTrigger/containers/CreateTrigger/utils/formikToTrigger';
import { triggerToFormik } from '../../../../CreateTrigger/containers/CreateTrigger/utils/triggerToFormik';
import { TRIGGER_TYPE } from '../../../../CreateTrigger/containers/CreateTrigger/utils/constants';
import { FORMIK_INITIAL_AGG_VALUES } from '../../CreateMonitor/utils/constants';
import { getInitialTriggerValues } from '../../../../CreateTrigger/components/AddTriggerButton/utils';

export const getInitialValues = ({
  title,
  index,
  timeField,
  flyoutMode,
  location,
  monitorToEdit,
  edit,
  searchType,
  detectorId,
}) => {
  let initialValues = _.mergeWith(
    {},
    _.cloneDeep(FORMIK_INITIAL_VALUES),
    initializeFromQueryParams(queryString.parse(location.search)),
    (initialValue, queryValue) => (_.isEmpty(queryValue) ? initialValue : queryValue)
  );

  if (flyoutMode) {
    initialValues.name = `${title} ${getDigitId()}`;
    initialValues.index = index;
    initialValues.timeField = timeField;

    // Add trigger
    const monitorType = initialValues.monitor_type;
    const initialTrigger = getInitialTriggerValues({ flyoutMode, monitorType, triggers: [] });
    initialValues.triggerDefinitions = [initialTrigger];

    // Add aggregations
    initialValues.aggregations = [_.cloneDeep(FORMIK_INITIAL_AGG_VALUES)];

    if (searchType) {
      initialValues.searchType = searchType;
    }

    if (detectorId) {
      initialValues.detectorId = detectorId;
    }
  }

  if (edit && monitorToEdit) {
    const triggers = triggerToFormik(_.get(monitorToEdit, 'triggers', []), monitorToEdit);
    initialValues = {
      ...monitorToFormik(monitorToEdit),
      triggerDefinitions: triggers.triggerDefinitions,
    };
  }

  return initialValues;
};

export const getPlugins = async (httpClient) => {
  try {
    const pluginsResponse = await httpClient.get('../api/alerting/_plugins');
    if (pluginsResponse.ok) {
      return pluginsResponse.resp.map((plugin) => plugin.component);
    } else {
      console.error('There was a problem getting plugins list');
      return [];
    }
  } catch (e) {
    console.error('There was a problem getting plugins list', e);
    return [];
  }
};

export const prepareTriggers = ({
  trigger,
  triggerMetadata,
  monitor,
  edit,
  triggerToEdit = [],
}) => {
  const { ui_metadata: uiMetadata = {}, triggers, monitor_type } = monitor;
  let updatedTriggers;
  let updatedUiMetadata;

  if (edit) {
    updatedTriggers = _.isArray(trigger) ? trigger.concat(triggers) : [trigger].concat(triggers);
    updatedUiMetadata = {
      ...uiMetadata,
      triggers: { ...uiMetadata.triggers, ...triggerMetadata },
    };
  } else {
    let updatedTriggersMetadata = _.cloneDeep(uiMetadata.triggers || {});

    let triggerType;
    switch (monitor_type) {
      case MONITOR_TYPE.BUCKET_LEVEL:
        triggerType = TRIGGER_TYPE.BUCKET_LEVEL;
        break;
      case MONITOR_TYPE.DOC_LEVEL:
        triggerType = TRIGGER_TYPE.DOC_LEVEL;
        break;
      default:
        triggerType = TRIGGER_TYPE.QUERY_LEVEL;
        break;
    }

    if (_.isArray(triggerToEdit)) {
      const names = triggerToEdit.map((entry) => _.get(entry, `${triggerType}.name`));
      names.forEach((name) => delete updatedTriggersMetadata[name]);
      updatedTriggers = _.cloneDeep(trigger);
    } else {
      const { name } = _.get(triggerToEdit, `${triggerType}`);
      delete updatedTriggersMetadata[name];

      const findTriggerName = (element) => {
        return name === _.get(element, `${triggerType}.name`);
      };

      const indexToUpdate = _.findIndex(triggers, findTriggerName);
      updatedTriggers = triggers.slice();
      updatedTriggers.splice(indexToUpdate, 1, trigger);
    }

    updatedUiMetadata = {
      ...uiMetadata,
      triggers: { ...updatedTriggersMetadata, ...triggerMetadata },
    };
  }

  return { triggers: updatedTriggers, ui_metadata: updatedUiMetadata };
};

export const create = async ({
  monitor,
  formikBag,
  httpClient,
  notifications,
  history,
  onSuccess,
}) => {
  const { setSubmitting } = formikBag;

  try {
    const resp = await httpClient.post('../api/alerting/monitors', {
      body: JSON.stringify(monitor),
    });
    setSubmitting(false);
    const {
      ok,
      resp: { _id },
    } = resp;
    if (ok) {
      notifications.toasts.addSuccess(`Monitor "${monitor.name}" successfully created.`);
      history.push(`/monitors/${_id}`);

      if (onSuccess) {
        onSuccess({ monitorId: _id });
      }
    } else {
      console.log('Failed to create:', resp);
      backendErrorNotification(notifications, 'create', 'monitor', resp.resp);
    }
  } catch (err) {
    console.error(err);
    setSubmitting(false);
  }
};

export const update = async ({ history, updateMonitor, notifications, monitor, formikBag }) => {
  const { setSubmitting } = formikBag;
  const updatedMonitor = _.cloneDeep(monitor);
  try {
    const resp = await updateMonitor(updatedMonitor);
    setSubmitting(false);
    const { ok, id } = resp;
    if (ok) {
      notifications.toasts.addSuccess(`Monitor "${monitor.name}" successfully updated.`);
      history.push(`/monitors/${id}`);
    } else {
      console.log('Failed to update:', resp);
    }
  } catch (err) {
    console.error(err);
    setSubmitting(false);
  }
};

export const submit = ({
  values,
  formikBag,
  edit,
  triggerToEdit,
  history,
  updateMonitor,
  notifications,
  httpClient,
  onSuccess,
}) => {
  let monitor = formikToMonitor(values);

  if (!_.isEmpty(_.get(values, 'triggerDefinitions'))) {
    const monitorUiMetadata = _.get(monitor, 'ui_metadata', {});
    const triggerMetadata = formikToTriggerUiMetadata(values, monitorUiMetadata);
    const triggers = prepareTriggers({
      trigger: formikToTrigger(values, monitorUiMetadata),
      triggerMetadata,
      monitor,
      edit,
      triggerToEdit,
    });
    monitor = { ...monitor, ...triggers };
  }

  if (edit) {
    update({ history, updateMonitor, notifications, monitor, formikBag });
  } else {
    create({ history, monitor, formikBag, httpClient, notifications, onSuccess });
  }
};
