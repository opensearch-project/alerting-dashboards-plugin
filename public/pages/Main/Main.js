/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Component } from 'react';
import { Switch, Route } from 'react-router-dom';
import { CoreConsumer } from '../../utils/CoreContext';

import Home from '../Home';
import Breadcrumbs from '../../components/Breadcrumbs';
import CreateMonitor from '../CreateMonitor';
import MonitorDetails from '../MonitorDetails/containers/MonitorDetails';
import CreateDestination from '../Destinations/containers/CreateDestination';
import Flyout from '../../components/Flyout';
import { APP_PATH } from '../../utils/constants';

class Main extends Component {
  state = { flyout: null };

  // TODO: Want to move this to redux store so we don't have to pass down setFlyout through components
  setFlyout = (flyout) => {
    const { flyout: currentFlyout } = this.state;
    // If current flyout and new flyout are same type, set to null to mimic closing flyout when clicking on same button
    if (currentFlyout && flyout && currentFlyout.type === flyout.type) {
      this.setState({ flyout: null });
    } else {
      this.setState({ flyout });
    }
  };

  render() {
    const { flyout } = this.state;
    const { history, ...rest } = this.props;
    return (
      <CoreConsumer>
        {(core) =>
          core && (
            <div style={{ padding: '15px 0px' }}>
              <Breadcrumbs history={history} httpClient={core.http} {...rest} />
              <Flyout
                flyout={flyout}
                onClose={() => {
                  this.setFlyout(null);
                }}
              />
              <Switch>
                <Route
                  path={APP_PATH.CREATE_MONITOR}
                  render={(props) => (
                    <CreateMonitor
                      httpClient={core.http}
                      setFlyout={this.setFlyout}
                      notifications={core.notifications}
                      isDarkMode={core.isDarkMode}
                      {...props}
                    />
                  )}
                />
                <Route
                  path={APP_PATH.CREATE_DESTINATION}
                  render={(props) => (
                    <CreateDestination
                      httpClient={core.http}
                      setFlyout={this.setFlyout}
                      notifications={core.notifications}
                      {...props}
                    />
                  )}
                />
                <Route
                  path="/destinations/:destinationId"
                  render={(props) => (
                    <CreateDestination
                      httpClient={core.http}
                      setFlyout={this.setFlyout}
                      notifications={core.notifications}
                      {...props}
                      edit
                    />
                  )}
                />
                <Route
                  path="/monitors/:monitorId"
                  render={(props) => (
                    <MonitorDetails
                      httpClient={core.http}
                      setFlyout={this.setFlyout}
                      notifications={core.notifications}
                      isDarkMode={core.isDarkMode}
                      {...props}
                    />
                  )}
                />
                <Route
                  render={(props) => (
                    <Home
                      httpClient={core.http}
                      {...props}
                      setFlyout={this.setFlyout}
                      notifications={core.notifications}
                    />
                  )}
                />
              </Switch>
            </div>
          )
        }
      </CoreConsumer>
    );
  }
}

export default Main;
