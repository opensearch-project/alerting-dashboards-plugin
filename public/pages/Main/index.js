/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import Main from './Main';

export default Main;
