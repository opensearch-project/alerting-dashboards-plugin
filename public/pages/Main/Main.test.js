/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { Router, Route, HashRouter } from 'react-router-dom';
import { render, mount } from 'enzyme';
import * as Breadcrumbs from '../../components/Breadcrumbs/Breadcrumbs';
import { createMemoryHistory } from 'history';

import Main from './Main';
import { setupCoreStart } from '../../../test/utils/helpers';

beforeAll(() => {
  setupCoreStart();
});

describe('Main', () => {
  test('renders', () => {
    const component = (
      <HashRouter>
        <Route render={(props) => <Main httpClient={{}} {...props} />} />
      </HashRouter>
    );

    expect(render(component)).toMatchSnapshot();
  });

  test('updates breadcrumbs when location updates', () => {
    const getBreadcrumbs = jest.spyOn(Breadcrumbs, 'getBreadcrumbs');
    const history = createMemoryHistory();
    history.push('/');
    mount(
      <Router history={history}>
        <Route render={(props) => <Main httpClient={{}} {...props} />} />
      </Router>
    );

    expect(getBreadcrumbs).toHaveBeenCalledTimes(1);
    history.push('/monitors');
    expect(getBreadcrumbs).toHaveBeenCalledTimes(2);
  });
});
