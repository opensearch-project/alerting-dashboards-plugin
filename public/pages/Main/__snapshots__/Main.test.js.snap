// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main renders 1`] = `
<div
  style="padding:15px 0px"
>
  <nav
    aria-label="breadcrumb"
    class="euiBreadcrumbs euiBreadcrumbs--truncate"
    style="padding:0px 15px"
  />
  <div>
    <div
      class="euiTabs"
      role="tablist"
    >
      <button
        aria-selected="true"
        class="euiTab euiTab-isSelected"
        role="tab"
        type="button"
      >
        <span
          class="euiTab__content"
        >
          Alerts
        </span>
      </button>
      <button
        aria-selected="false"
        class="euiTab"
        role="tab"
        type="button"
      >
        <span
          class="euiTab__content"
        >
          Monitors
        </span>
      </button>
      <button
        aria-selected="false"
        class="euiTab"
        role="tab"
        type="button"
      >
        <span
          class="euiTab__content"
        >
          Destinations
        </span>
      </button>
    </div>
    <div
      style="padding:25px 25px"
    />
  </div>
</div>
`;
